### PR TITLE
fix #8862: cannot watch unicode properties like 'a.中文'

### LINKED ISSUE #8862: cannot watch unicode properties like  'a.中文'
### Version
2.5.17

### Reproduction link
https://jsfiddle.net/k72f3Lon/

### Steps to reproduce
1.webpack project with vue .
2.add a component .
3.create instance of Vue .
4.add a property `data:function(){return {a:{中文:"",b:""}}}` to the instance.
5.add watch properties to the instance as `watch:{'a.b':function(val){},'a.中文':function(val){}}`
6.console show error that Watcher only accepts simple dot-delimited paths. 

### What is expected?
no error the code run continue

### What is actually happening?
the code seems stopped to run

### PR DESCRIPTION
**Pull request by derek-li**
Hi all at Vue!

This pull request resolves the issue regarding watchers not watching unicode properties like 'a.中文'.

My method for resolving this case is to throw a warning when non-alphanumeric, $, or _ characters are detected. My reasoning behind throwing a warning instead of supporting special characters is that I think  supporting special characters would promote poor naming habits that could only make code more difficult to read and understand.

I'm pretty new to open source so any feedback would be awesome and please let me know if you'd like to see any changes or improvements!

Link to issue: https://github.com/vuejs/vue/issues/8862

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

### CODE REVIEW

**Comment by posva** [comment]
This is not fixing the issue it says to be fixing 🤔

File: src/core/observer/watcher.js [hunk 1]
```diff
@@ -80,13 +80,21 @@ export default class Watcher {
     } else {
       this.getter = parsePath(expOrFn)
       if (!this.getter) {
+        if (this.getter === false) {
```
Comment (yyx990803) [inline src/core/observer/watcher.js]: I don't like that we are relying on magic return values (`false`) here, as the logic is separated in two places. Also the check is warning only so it's better to place it right here inside a `if (process.env.NODE_ENV !== 'production') { ... }` block, using an inline regexp.

**Comment by yyx990803** [comment]
Note this may become invalid if we merge #8666, I'm still debating whether we should support it.

File: src/core/observer/watcher.js (hunk 1 above)
Comment (derek-li) [reply src/core/observer/watcher.js]: @yyx990803 Thanks for taking the time to review this PR. 

I've gone ahead and made the commits to stop relying on random return values. I also tried your suggestion with having the check in an `if (process.env.NODE_ENV !== 'production') { ... }` block, but felt that it was unclear to have two regex checks in different places (`watcher` and `parsePath`) and did not resolve your comment on the logic being separated, so I grouped them together.

Thanks again and please let me know what you think!

**Comment by yyx990803** [comment]
Closed via #8666

**Pull request closed without merging**